File: src/main/java/com/workouts/myworkouts/service/training/TrainingService.java
package com.workouts.myworkouts.service.training;


import com.workouts.myworkouts.model.dto.training.TrainingExerciseDto;
import com.workouts.myworkouts.model.dto.training.TrainingPlanDto;

import java.time.LocalDate;
import java.util.List;

public interface TrainingService {

    TrainingPlanDto createTraining(TrainingPlanDto trainingDTO);

    List<TrainingPlanDto> getAllTrainings();

    TrainingPlanDto getTrainingWithFetchedWorkoutExercise(long id);

    TrainingPlanDto addExerciseToTraining(long trainingId, TrainingExerciseDto trainingExerciseDto);

    boolean applyTraining(long trainingId, LocalDate dateTime);

    void resetTrainings();

    void deleteTraining(long trainingId);

    TrainingPlanDto updateTraining(long trainingId, TrainingPlanDto trainingDTO);

    boolean setScheduledForTraining(long trainingId, boolean scheduled);

}

File: frontend/src/app/services/rest/training/training.service.ts
import {Injectable} from '@angular/core';
import {Observable} from "rxjs";
import {HttpClient, HttpParams} from "@angular/common/http";
import {Urls} from "../../../model/urls";
import {TrainingPlan} from "../../../model/training/trainingPlan";
import {TrainingExercise} from "../../../model/training/trainingExercise";
import moment from "moment";

@Injectable({
  providedIn: 'root'
})
export class TrainingService {

  constructor(private http: HttpClient) { }

  public createTraining(training: TrainingPlan): Observable<TrainingPlan> {
    return this.http.post<TrainingPlan>(Urls.API_URL + "/trainings", training);
  }

  public setScheduledForTraining(trainingId: number, scheduled: boolean): Observable<boolean> {
    return this.http.get<boolean>(Urls.API_URL + "/trainings/" +  trainingId + "/setScheduled", {params: new HttpParams().set("scheduled", scheduled.toString())});
  }

  public updateTraining(trainingId: number, training: TrainingPlan): Observable<TrainingPlan> {
    return this.http.post<TrainingPlan>(Urls.API_URL + "/trainings/" +  trainingId, training);
  }

  public getAllTrainingsWithoutWorkouts(): Observable<TrainingPlan[]> {
    return this.http.get<TrainingPlan[]>(Urls.API_URL + "/trainings");
  }

  public addExerciseToTraining(trainingId: number, trainingExercise: TrainingExercise): Observable<TrainingPlan> {
    return this.http.post<TrainingPlan>(Urls.API_URL + "/trainings/" + trainingId + "/exercises", trainingExercise);
  }

  public applyTraining(trainingId: number, dateTime: Date): Observable<boolean> {
    const localDate = moment(dateTime).format('YYYY-MM-DD') + 'T00:00:00';
    return this.http.get<boolean>(Urls.API_URL + "/trainings/" + trainingId + "/" + localDate);
  }

  public deleteTraining(trainingId: number): Observable<boolean> {
    return this.http.delete<boolean>(Urls.API_URL + "/trainings/" + trainingId);
  }

  public getTrainingWithWorkouts(trainingId: number): Observable<TrainingPlan> {
    return this.http.get<TrainingPlan>(Urls.API_URL + "/trainings/" + trainingId);
  }
}
